### PR TITLE
Restructure compatibility testing

### PR DESCRIPTION
**Pull request by unicod3**
Resolves https://github.com/open-telemetry/opentelemetry-go/issues/3768

Hello, this is my first contribution to this repo so tried to tackle this good first issue. 

I've restructured the compatibility tests accordingly to what was discussed in last week's meeting (2023-02-23).

Also renamed the token job's name to `test-compatibility` so it is aligned with other test jobs.

You can see it in action in the link below:
https://github.com/unicod3/opentelemetry-go/actions/runs/4272402129/jobs/7437568975

Please let me know if I missed anything.

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -112,7 +112,7 @@ jobs:
           name: opentelemetry-go-test-output
           path: ${{ env.TEST_RESULTS }}
 
-  compatibility-test:
+  compatibility-matrix:
```
Comment (MrAlias) [inline .github/workflows/ci.yml:115]: Changing this name means the existing required check will not be run. The original name should be kept here and if we want to change it we can do so once the PR required checks are moved to the require test-compatibility.

Comment (unicod3) [reply .github/workflows/ci.yml]: oh good catch, I reverted it back to the original name.

**Comment by dmathieu** [comment]
I'm not sure this requires a changelog entry.

**Comment by unicod3** [comment]
> I'm not sure this requires a changelog entry.

Agreed, I wasn't sure either tbh, removed it from PR.

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -150,3 +150,19 @@ jobs:
...
+        run: echo ${{ needs.compatibility-test.result }}
+      - name: Interpret result
+        run: |
+          if [[ success == ${{ needs.compatibility-test.result }} ]]
+          then
+            echo "All matrix jobs passed!"
+          else
+            echo "One or more matrix jobs failed."
+            false
+          fi
```
Comment (pellared) [inline .github/workflows/ci.yml:168]: nit: Why do we have two steps?

```suggestion
      - name: Test if compatibility-test passed
        run: |
          echo "${{ needs.compatibility-test.result }}"
          test "${{ needs.compatibility-test.result" }} == "success"
```

**Review by pellared — APPROVED** [review]
nit comment (with proposed changed)

Thanks for contribution.

File: .github/workflows/ci.yml (hunk 2 above)
Comment (unicod3) [reply .github/workflows/ci.yml]: Thanks for the suggestion, quotes were not aligned with the suggestion, so did a local commit and updated the PR.

**Review by MrAlias — APPROVED** [review]

**Pull request merged**